src/protocols/MeterModbus.cpp

```diff
@@ -0,0 +1,213 @@
+
+	_mbconn.reset(new ModbusRTUConnection(rtu_device, rtu_baudrate));
+}
+
+void MeterModbus::create_tcp(const std::list<Option> &options) {
+	std::string tcp_host;
+	int tcp_port;
+	tcp_host = lookup_mandatory<const char *, OptionList::lookup_string>(options, "host", name());
+	tcp_port = lookup_optional<int, OptionList::lookup_int>(options, "port", 502);
+
+	_mbconn.reset(new ModbusTCPConnection(tcp_host, tcp_port));
+}
+
+MeterModbus::~MeterModbus()
+{
+	// TODO Auto-generated destructor stub
+}
+
+int MeterModbus::open() {
+	try {
+		_mbconn->connect();
+	} catch (ModbusException &e) {
+		print(log_error, "MeterModbus can not connect: %s. Will retry.", name().c_str(), e.what());
+	}
+	if (_libmodbus_debug)
+		_mbconn->debug(true);
+	return SUCCESS;
+}
+
+int MeterModbus::close() {
+	_mbconn.reset();
+	return SUCCESS;
+}
+
+ssize_t MeterModbus::read(std::vector<Reading> &rds, size_t n) {
+	bool one_success = false;
+	rds.clear();
+	for (auto&& i : _devices) {
+		slaveid_t slave = i.first;
+		RegisterMap::Ptr regmap = i.second;
+		try {
+			regmap->read(rds, _mbconn, slave);
+			one_success = true;
+		} catch (ModbusException &e) {
+			print(log_error, "Modbus read error slave %d: %s. Skip and re-connect", name().c_str(), slave, e.what());
+			try {
+				_mbconn->close();
+				_mbconn->connect();
+			} catch (ModbusException &e) {
+				print(log_error, "Reconnect failed: %s. Will retry.", name().c_str(), e.what());
+			}
+		}
+	}
+	if (!one_success) {
+		print(log_error, "Modbus read unsuccessful.", name().c_str());
+	}
+	return rds.size();
+}
+
+void ModbusConnection::connect() {
+	if (!_ctx)
+		open();
+	int ret = modbus_connect(_ctx);
+	if (ret < 0) {
+		throw ModbusException("connecting modbus");
+	}
+	_connected = true;
+
+}
+
+void ModbusConnection::close() {
+	if (_ctx) {
+		if (_connected) {
+			modbus_close(_ctx);
+			_connected = false;
+		}
+		modbus_free(_ctx);
+		_ctx = NULL;
+	}
+}
+
+ModbusConnection::~ModbusConnection() {
+	close();
+}
+
+void ModbusConnection::read_registers(int addr, int nb, uint16_t *dest, unsigned slave) {
+	if (!_ctx || !_connected)
+		throw(ModbusException("modbus_read_registers not connected."));
+	modbus_set_slave(_ctx, slave);
+	int ret = modbus_read_registers(_ctx, addr, nb, dest);
+	if (ret < 0)
+		throw(ModbusException("modbus_read_registers failed."));
+}
+
+void ModbusConnection::debug(bool enable) {
+	modbus_set_debug(_ctx, TRUE);
+}
+void ModbusRTUConnection::open() {
+	close();
+	_ctx = modbus_new_rtu(_device.c_str(), _baud, 'N', 8, 1);
+	if (_ctx == NULL)
+		throw ModbusException("creating new rtu");
+}
+
+void ModbusTCPConnection::open() {
+	close();
+	_ctx = modbus_new_tcp(_ip.c_str(), _port);
+	if (_ctx == NULL)
+		throw ModbusException("creating new tcp");
+}
+
+
+std::map<std::string, RegisterMap::Ptr (*)()> RegisterMap::maps = {
+		{ "ohmpilot", createMap<OpRegisterMap> },
+		{ "imemeter", createMap<IMEmeterRegisterMap> }
+};
```

Comment:
Oh. Oooooh...
Nice to have an "easy way" to adresse known type of meters, but:
This is inflexible! For every new meter or single parameter we need somebody to understand and change the code and users have to compile again.